pkg/ha/ha_parser.go

```diff
@@ -28,31 +28,34 @@ func (h *haParser) ParseData(tts []prompb.TimeSeries) (map[string][]model.Sample
 	dataSamples := make(map[string][]model.SamplesInfo)
 	rows := 0
 
-	var replicaName, clusterName string
-	if len(tts) > 0 {
-		s, _, err := model.LabelProtosToLabels(tts[0].Labels)
-		if err != nil {
-			return nil, rows, err
-		}
-		replicaName = s.GetReplicaName()
-		clusterName = s.GetClusterName()
+	if len(tts) == 0 {
+		return dataSamples, rows, nil
+	}
+
+	s, _, err := model.LabelProtosToLabels(tts[0].Labels)
+	if err != nil {
+		return nil, rows, err
 	}
+	replicaName := s.GetReplicaName()
+	clusterName := s.GetClusterName()
 
 	if err := checkClusterAndReplicaLabelAreSet(clusterName, replicaName); err != nil {
 		return nil, rows, err
 	}
 
 	// find samples time range
 	var minTUnix, maxTUnix int64
+	minTWasSet := false
 	for i := range tts {
```

Comment:
Do you mean the loop here and in line 80?

Reply:
yes, trying to make a single looping over `tts`.

Reply:
it's done on purpose. the second loop also 1) parses all the labels, 2) filters through only accepted time range, and 3) populates the samples per metric map. 1 and 3 -> not need to do it at all if checkInsert fails, 2 -> have to do it in 2 places since checkInsert tells you the accepted time range